Dont use a null value when unable to parse variable

rfvgyhn opened the pull request:
So I ran into a save file that has a variable with the magic number `PO*P` instead of `PORP`. When the un-parsable exception is caught it never sets a value to the `variables` array which then causes a null reference exception in `SavegameFile.ReferenceVariable` when trying to access `nextVariable.TokenSize`.

I don't know if `PO*P` variables are supposed to be different from `PORP` variables or if falling back to `UnknownVariable` would be more preferable than the `InvalidVariable` I introduced. Let me know what you think.

[ManualSave_107631_7e400400_535c8ec.zip](https://github.com/Atvaark/W3SavegameEditor/files/4017934/ManualSave_107631_7e400400_535c8ec.zip)

rfvgyhn committed:
dont use a null value when unable to parse variable

Atvaark commented:
Looks good to me. I don't know any more about this than you.

Have you checked if these unknown variables can be read with the PorpVariableParser?
You'd have to update this check:
https://github.com/Atvaark/W3SavegameEditor/blob/762738be0814530a49aa5ea3aa050b907e0c087f/W3SavegameEditor.Core/Savegame/VariableParsers/PorpVariableParser.cs#L46-L48

rfvgyhn commented:
It can't be read by the porp parser. It's actually a 4 byte token whos name ends up matching with the `W` component of a vector. It looks like its an unknown variable before (`3}??`) and after (`  ??`) it though so I'm wondering if this is also just an unknown var that happens to have the same two byte magic number as `PORP`. Perhaps this value got in there from a modded game?

rfvgyhn commented:
Another way to go about it is to not do a partial magic string match. Could do a `PeekString(2)` and if that doesn't return a match do a `PeekString(4)` (after updating the parsers to return their full magic strings). Then it could just fallback on the existing `UnknownVariable` logic and not introduce a new `InvalidVariable` type.

https://github.com/Atvaark/W3SavegameEditor/blob/762738be0814530a49aa5ea3aa050b907e0c087f/W3SavegameEditor.Core/Savegame/VariableParser.cs#L43-L45

rfvgyhn committed:
Use full magic strings when parsing instead of first two chars

Helps prevent parse errors when an unknown variable looks like a known var (e.g. AV?C uses AVAL parser)

rfvgyhn committed:
Add stricter verification for BS and SS var parsers

Handles unknown variables that look like SS or BS vars (e.g. SS{C and BSF?)

rfvgyhn committed:
Don't use a null value when able to parse variable

rfvgyhn commented:
I was provided with another save file and it turns out there are more look-alike variables. Attached save has variables with the following magic strings: `AV?C`, `BSF?` and `SS{C`. None of which are parsable by the existing parsers.

I've update the PR to try and match the full magic strings instead of just the first two chars. I also included some verification for BS and SS vars for when it encounters the look-alikes.

[ManualSave_52587_7e404800_40689b5.zip](https://github.com/Atvaark/W3SavegameEditor/files/4088913/ManualSave_52587_7e404800_40689b5.zip)

Atvaark commented:
Looks good to me.
